### ISSUE TITLE
reference key as target-finder?

### ISSUE THREAD

**Issue by bdarcus** [pprevos/citar-denote#13]
Was playing a bit with your package and denote, and noticed something that sometimes comes up in the org-roam world.

The `reference` property citekey: you may want to consider an embark "target-finder" (or similar for hydra/transient) for that, so that users can act on it at point.

If you conclude yes, you may want to consider a syntax with already-existing finder: `@key`?

**Comment by pprevos** [comment]
Hi Bruce,

In the version currently on my workbench I wrote citar-denote-dwim which calls the citar-dwm function for the citations keys in the reference line. 

(defun citar-denote-dwim ()
  "Access attachments, notes and links associated with a bibliographic note."
  (interactive)
  ;; Any citation keys in the note?
  (if-let ((keys (citar-denote-retrieve-references (buffer-file-name))))
      ;; Check if citation keys are in the bibliography
      (if-let (keys? (not (seq-every-p 'null
                                       (mapcar (lambda (key)
                                                 (gethash key (citar-get-entries))) keys))))
          (citar-open keys)
        (user-error "Citation key(s) not in bibliography"))
    (if (denote-file-is-note-p (buffer-file-name))
        (when (yes-or-no-p "Current buffer does not reference a citation key.  Add a reference ?")
          (citar-denote-add-citekey)
          (citar-denote-dwim))
      (user-error "Buffer is not a Denote file"))))

Are these the sort of actions you seek?

Wold introducing the ampersand create backward compatibility issues?

Wrapping the references in properties is not in line with the Denote way of working.

**Comment by bdarcus** [comment]
> Are these the sort of actions you seek?

I'm not really requesting it per se; more just noting what others have requested in related packages.

In `citar`, if you have point on a key in a property drawer, and you run `embark-act`, it will present you with a list of actions you can run on it. [This transient menu](https://github.com/emacs-citar/citar/wiki/Transient-org-cite-follow-processor) does the same.

> Would introducing the ampersand create backward compatibility issues?

IDK.

> Wrapping the references in properties is not in line with the Denote way of working.

So rather than using standard org properties for metadata, the Denote "way" is to use non-standard equivalents (I see now they're actually not org properties because the syntax is slightly different), but not specific to org?

**Comment by bdarcus** [comment]
You mean ampersand (&) or asperand (@)?

> Using a property line is normal practice in Org Mode so we are not deviating from standard Org Mode practice.

The properties that denote or citar-denote (I don't understand which) creates now aren't actually properties, so org doesn't see them. That's why that function returns nil, even if I add the asperand to the key.

https://orgmode.org/guide/Properties.html

**Comment by pprevos** [comment]
Denote works this way to create consistency with Markdown and text files. It does not aim to be Org Mode centric. 

It seems that `(org-element-parse-buffer 'greater-element)` recognises the Denote in-buffer settings, i.e. identifier and reference.

Using `(org-element-at-point)` on the reference line gives the correct value. 

This page has some good info: https://orgmode.org/worg/dev/org-element-api.html

I don't have any time to work on this until May, but I think we could make the Embark at point function work on the reference line.

**Comment by bdarcus** [comment]
Yeah, I may just be not understanding some details about denote etc. 

In the end, it should ideally just be easy for the current citar at-point functionality to work well with both org and markdown (via the yaml or toml headers).

Right now, citar:

- IIRC, actually only supports org
- does not assume org-roam per se (it doesn't look for the `ROAM_REF` key), but does assume a metadata citekey value of the form `@citekey` (which org-roam supports, but doesn't require)

So there be room to collaborate there when the time comes.

**Comment by pprevos** [comment]
Thanks. I'll review all this in May (big trip to the US coming up). Ideally we get Citar to integrate with Denote using Org, Markdown and plain text.

**Comment by pprevos** [comment]
> Aside: seems the new [djot](https://djot.net/) format is likely to adopt the same YAML-light metadata syntax as pandoc, and a very similar citation syntax. So at the inevitable point at which we see an Emacs note package that supports that format, citar already does!

It is easy to configure Denote to accept djot by configuring `denote-file-types`. There does not seem to be a major mode for this format yet.

**Comment by bdarcus** [comment]
Can any of you test this PR and confirm it works as expected?

https://github.com/emacs-citar/citar/pull/732

**Comment by pprevos** [comment]
Hi Bruce, I wrote that I don't have time to look at this issue till May (which is true), but than I got distracted by the other issue. Anyway, I should be doing other things, but the allure of Emacs hacking is strong.

**Comment by pprevos** [comment]
I thought a bit about this one. The current `citar-denote-dwim` function lets the user access resources related to references in a note by selecting  a reference (if more than one) and umping to `citar-open`. This allows the user to get context on references without having to have point on it.

I could redirect this function to the Embark equivalent, in case `citar-embark-mode` is enabled. Would that be a acceptable solution to this issue?

**Comment by bdarcus** [comment]
IDK.

I can just say that from a user POV, in `citar` and `citar-org-roam`, at-point behavior is the same with citekey in the buffer, and in the metadata properties, regardless of what function one is using at-point (embark, a transient menu or hydra, etc.; the only exception is `org-open-at-point` won't recognize the reference in the property drawer).
